강원대 BE_서영진 2단계 - 카카오 주문하기

seoyoungjin23 opened the pull request:
## 미션 핵심 기능 구현 사항
1. 카카오 주문하기 시 상품 옵션의 수량 차감
2. 카카오 주문하기 시 해당 상품이 위시 리스트에 있으면 위시리스트에서 삭제
3. 카카오 주문하기 시 주문 상품, 수량, 메세지를 담아서 카카오톡 메세지 전송
<img width="345" height="272" alt="image" src="https://github.com/user-attachments/assets/904f0569-f628-4abb-8277-983cedf99fb3" />

## 주요 기능 구현 사항
1. **카카오 서버 `accessToken`, `refreshToken` DB 저장 및 재발급 로직 구현**  
   - 두 토큰 모두 DB에 저장하여 관리합니다!
   - `@CurrentUser` 리졸버를 통해 토큰을 쉽게 사용할 수 있도록 구현하였습니다.  
   - 토큰 만료 시 DB에 저장된 `refreshToken`을 통해 자동 재발급 처리하도록 로직을 구현하였습니다.

2. **기존 JSON 응답 → 쿠키 방식으로 변경**  
   - 기존에는 토큰을 JSON으로 반환했으나, 쿠키를 반환하는 방식으로 수정했습니다!

## 주요 리팩토링 사항
1. **상품 수정 로직 개선**  
   - 기존 상품 수정에서 옵션까지 바꾸는 교체 방식에서 → 옵션 개별 수정 방식으로 변경했습니다!

2. **도메인 VO 생성 및 검증 로직 분리**  
   - 제가 느끼기에 검증 로직이 엔티티에 섞여 있는 것이 보기에 좀 불편하기도 하였고,  
     큰 프로젝트에서 필드가 많은 도메인의 경우 더 많은 검증 로직이 존재할 때 더 불편할 것 같아서  
     이를 어떻게 분리할까 알아보다가 VO 라는 개념을 알게되어서 적용해보았습니다!  
   - VO 생성으로 검증 로직을 VO 내부로 이동시켜 책임을 분리해보았습니다!

3. **카카오 API 클라이언트 모듈화**  
   - 카카오의 외부 API를 요청하는 기능들을 하나의 모듈로 만들어서 사용하면 좋을 것 같아서`KaKaoMessageClient`, `KaKaoTokenClient` 로 별도의 빈을 만들어서 모듈화해보았습니다!

4. **패키지 구조 개선**  
   - 계속 클래스가 늘어날수록 파일을 찾는 데 시간이 너무 걸리고 가독성이 떨어져서 도메인 중심 패키지 구조로 변경해보았습니다!

5. **예외 처리 개선**  
   - `CustomException`에 `HttpStatus` 기반 예외 코드를 추가하고, 글로벌 핸들러에도 적용시켰습니다!


### 사용 방법
1. 이전 PR https://github.com/next-step/spring-gift-order/pull/111 에서의 로그인 방식을 활용하여 카카오 로그인을 합니다.
2. `Post http://localhost:8080/api/orders` 다음과 같은 요청을 보내면 응답 및 카카오톡 메세지를 받을 수 있습니다!

- 응답 예시

  **Request**

  ```json
  {
    "optionId": 1,
    "quantity": 1,
    "message": "생일 선물 ~"
  }
  ```

  **Response**

  ```json
  {
    "status": 201,
    "message": "생성 성공",
    "data": {
      "id": 1,
      "productId": 1,
      "optionId": 1,
      "quantity": 1,
      "message": "생일 선물 ~"
    }
  }
  ```

## 궁금한 사항
1. **검증의 책임과 관련하여서**
  - **고민하게된 배경**
    현재 저는 유효성 검증을 다음과 같이 하고 있습니다
     - DTO: 클라이언트로부터 받은 데이터를 받을 때 기본적인 형식/유효성 검증 수행
     - 엔티티: 내부에서 VO를 통해 다시 한 번 더 유효성 검증 + 비즈니스  로직 검증
    처음엔 안전성을 위해 이렇게 1차-2차 검증 구조를 선택했습니다.

  - **문제 의식 및 고민**
    - 검증 로직을 양쪽(DTO, VO)에 중복해서 작성하는 것이 개발 리소스를 낭비하는 건 아닐까? 라는 생각을 하게되었습니다.
    - 변경이 가능한 필드라면 VO에서의 검증은 당연히 필요하다고 생각합니다.
    - 하지만 단순히 클라이언트로부터 받은 값을 그대로 사용하는 경우라면, DTO에서만 검증하고 VO에서는 유효성 검증을 생략해도 괜찮지 않을까? 라는 생각을 하게되었습니다.

  - **정리**
    - 제가 구현한 DTO와 VO 양쪽에서의 이중으로 유효성 검증이 적합한 방식인가요..? 
    - DTO에서 유효성, 엔티티에서는 비즈니스 로직만 검증에 집중하고 값의 변경이 있는 필드만 유효성 검증을 한 번 더 하면 되는 것일까요?
    - 멘토님이라면 검증 책임을 어떻게 두시는 지 궁금합니다!

seoyoungjin23 committed:
feat(schema): Member 엔티티에 accessToken 및 refreshToken 필드 추가

seoyoungjin23 committed:
feat(schema): Orders 엔티티 추가

seoyoungjin23 committed:
feat(auth): 카카오 로그인 시 accessToken과 refreshToken 저장 로직 구현

seoyoungjin23 committed:
test(auth): accessToken 및 refreshToken 저장 로직에 따른 테스트 수정

seoyoungjin23 committed:
feat(auth): @CurrentUser에 accessToken, refreshToken이 포함되도록 변경

seoyoungjin23 committed:
chore(properties): 카카오 메시지 전송 URL 및 scope 설정 추가

seoyoungjin23 committed:
feat(order): 상품 주문 처리 로직 구현

- 주문 시 ProductOption의 수량 감소
- 주문된 상품이 위시리스트에 있을 경우 해당 항목 삭제

seoyoungjin23 committed:
feat(order): 카카오 메시지 전송 로직 추가

seoyoungjin23 committed:
feat(auth): 카카오 액세스 토큰, 리프레시 토큰 재발급 및 DB 업데이트 로직 추가

seoyoungjin23 committed:
feat(order): 주문 메시지 전송 시 accessToken 만료에 따른 재발급 및 재시도 로직 추가

seoyoungjin23 committed:
feat(auth): 카카오 토큰 요청 로직 KaKaoTokenClient로 분리

seoyoungjin23 committed:
test(wish): CurrentUser 리졸버의 인증 성공을 위한 MemberRepository mocking 추가

seoyoungjin23 committed:
feat(auth): 로그인 콜백에서 인증 토큰을 쿠키로 전달하도록 구현

seoyoungjin23 committed:
test(order): 주문 생성, 실패 케이스 등 E2E 테스트 작성

seoyoungjin23 committed:
test(auth): 카카오 콜백 요청 시 액세스 토큰 쿠키 설정 테스트 수정

seoyoungjin23 committed:
refactor: 서비스 로직에 @Transactional 어노테이션 추가

seoyoungjin23 committed:
fix(product): 상품 수정 시 옵션 수정 로직 제거

seoyoungjin23 committed:
feat(option): 옵션 개별 수정 기능 추가

seoyoungjin23 committed:
test: 옵션 수정 기능에 대한 테스트 코드 수정

seoyoungjin23 committed:
refactor: 패키지 구조 정리 및 도메인별 디렉터리 분리

seoyoungjin23 committed:
refactor: 중복된 Member 조회 로직 제거

seoyoungjin23 committed:
refactor: getter 메서드를 카멜케이스로 변경

seoyoungjin23 committed:
refactor: KaKaoTokenInfo에 JsonNaming 적용

seoyoungjin23 committed:
feat: CustomException에 HttpStatus 기반 예외 처리 방식 추가 및 핸들러 개선

seoyoungjin23 committed:
feat: 도메인 VO 생성 및 도메인 검증 로직 분리

seoyoungjin23 committed:
test: 예외 처리 로직 변경에 따른 테스트 코드 수정

seoyoungjin23 committed:
refactor: 중복된 검증 및 사용하지 않는 코드 제거

seoyoungjin23 committed:
feat: wish VO 생성 및 도메인 검증 로직 분리

seoyoungjin23 committed:
feat: 접근 제어자 수정 및 생성자 수정

seoyoungjin23 committed:
refactor: 도메인 패키지 구조 정리

seoyoungjin23 committed:
refactor: 연관관계 중복 설정 문제 해결

seoyoungjin23 committed:
refactor: CustomException에서 ValidationException으로 명확한 예외 구분

seoyoungjin23 committed:
readme(order): 카카오 주문하기 구현 readme 작성

urarik commented on src/main/java/gift/auth/cookie/CookieUtil.java:
```diff
@@ -0,0 +1,31 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class CookieUtil {
```
이 클래스를 여러 클래스가 사용하나요? 그렇지 않다면 이 클래스의 기능을 이렇게 분리하면서 얻는 이점은 어떤게 있을까요~?

seoyoungjin23 replied:
- 현재 CookieUtil 클래스가 여러 클래스에서 사용되고 있다면 유틸 클래스로 분리하는 것은 중복 제거와 재사용성 측면에서 이점이 있겠지만 하나의 클래스에서 사용되고 있습니다..

- 제가 생각하는 분리의 이점은 재사용성 측면과 중복 제거 정도라고 생각합니다! 하나의 클래스에서 사용하는데 분리하여 얻는 이점이 잘 생각나질 않습니다

- 그렇다면 굳이 유틸 클래스로 분리한 것이 오버엔지니어링 되었다고 생각하면 될까요? 이렇게 사용하는 범위가 한정적이라면 그냥 private 메서드로 만들고 재사용이 필요할 때 그때 컴포넌트로 분리하는 것이 오히려 더 좋은 구조일까요?

urarik replied:
@seoyoungjin23 제 생각에 이정도는 충분히 작성될 수 있는 클래스라고 생각합니다~ 객체지향의 근본은 High Cohesion, Low Coupling 이라고 생각하는데요. 이 원칙을 공부해보면 클래스를 나누는 근거를 어느정도 정립할 수 있을거예요

urarik commented on src/main/java/gift/auth/resolver/CurrentUserArgumentResolver.java:
```diff
@@ -25,12 +30,9 @@ public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer m
 
-        return member;
+        return memberRepository.findById(memberId)
+            .orElseThrow(() -> new CustomException(CustomResponseCode.UNAUTHORIZED));
```
예외 이름으로 예외를 잘 나타낼 수 있으면 더 좋지 않을까요?
지금은 new CustomException(CustomResponseCode.UNAUTHORIZED) 에서 의미있는 단어는 UNAUTHORIZED 뿐인 것 같네요~

seoyoungjin23 replied:
수정한다는 것을 깜빡하였습니다! 전체적으로 CustomException을 에러를 더 잘 표현할 수 있는 방법으로 수정하겠습니다!

urarik commented on src/main/java/gift/common/exception/core/CustomException.java:
```diff
@@ -0,0 +1,37 @@
+
+public class CustomException extends RuntimeException {
+
+    private final CustomResponseCode errorCode;
```
예외 클래스가 응답 코드도 알아야 할까요? 응답 코드는 어느 계층 혹은 모듈의 책임일까요?

seoyoungjin23 replied:
- 말씀 주신 대로 예외 클래스에서 응답 코드를 직접 포함하는 것은 지양하는 것이 맞다고 생각합니다! 
- 응답 코드는 결국 클라이언트와의 통신에서만 의미가 있는 값이기 때문에 이러한 책임은 컨트롤러나 글로벌 예외 핸들러에서 담당하는 것이 더 적절하다고 생각합니다! 리팩토링 하겠습니다!

urarik commented on src/main/java/gift/controller/api/OptionController.java:
```diff
@@ -48,4 +49,16 @@ public ResponseEntity<CustomResponseBody<List<ProductOptionResponse>>> getOption
+    ) {
+        ProductOptionResponse response = optionService.update(optionId, request);
+
+        return ResponseEntity
+            .status(CustomResponseCode.UPDATED.getHttpStatus())
+            .body(CustomResponseBody.of(CustomResponseCode.UPDATED, response));
```
이 코드를 처음 보면 UPDATED라는 http 응답 코드가 있는 것처럼 보여요. 그냥 ResponseEntity.ok를 쓰는게 훨씬 더 깔끔하지 않을까요? 
문제 해결에서 대부분 가장 좋은 방법은 널리 알려져있고 표준화된 방법을 사용하는 것입니다. 제 생각에 이런 식의 재정의는 나중에 혼란을 줄뿐 아니라 유지보수도 힘들 것 같아요.

seoyoungjin23 replied:
- 처음 설계할 때는 응답을 좀 더 명확하게 표현해주면 좋겠다는 생각이 있었습니다. 하지만 멘토님의 피드백을 보니, 사실은 굳이 쉬운 일을 두 번 하고 있었다는 생각이 듭니다..

- `.ok()`, `.noContent()`, 혹은 `.status()` 같은 표준화된 방식을 사용하는 게 누구나 바로 이해할 수 있어서 개발 리소스를 절약할 수 있을 것 같습니다!

urarik commented on src/main/java/gift/entity/Member/Member.java:
```diff
@@ -28,23 +28,47 @@ public class Member {
+    public Member(Long providerId, String email, String nickname, String profileImage,
+        String accessToken, String refreshToken) {
+        this(null, providerId, email, nickname, profileImage, accessToken, refreshToken);
+    }
```
생성자가 너무 많아지면 가독성도 안좋고 귀찮아져요. 빌더 패턴을 써보면 어떨까요?
Lombok의 빌더 어노테이션을 사용하면 간단하지만 지금은 lombok을 못 쓰는 것으로 알아서 한 번 직접 구현해봐도 좋을 것 같습니다~

urarik commented on src/main/java/gift/entity/Order/Order.java:
```diff
@@ -0,0 +1,72 @@
+    @JoinColumn(name = "product_option_id", nullable = false)
+    private ProductOption productOption;
+
+    @Embedded
+    private OrderQuantity quantity;
+
+    @Embedded
+    private OrderMessage message;
```
이렇게 필드가 하나 뿐인데 클래스를 각각 따로 만들면 장단점이 뭘까요?

seoyoungjin23 replied:
- 제가 느낀 장점은 quantity, name 같은 vo로 분리하면 해당 필드의 비즈니스 규칙을 한 곳에 모아서 관리할 수 있다는 점이 좋았습니다. 애초에 처음 만든 목적도 그러했습니다! 

- 하지만 직접 코드를 작성하면서 느낀 점은 필드가 하나 뿐인데 이렇게 까지 작성해야하나? 였습니다. 각 필드를 별도의 클래스로 분리하다보니 지나치게 클래스 수가 많아져서 처음 코드를 보았을 때 구조가 좀 난해하다 라는 느낌을 받았습니다!

urarik commented on src/main/java/gift/entity/Product/Option/ProductOptionQuantity.java:
```diff
@@ -0,0 +1,32 @@
+
+    public ProductOptionQuantity decrease(long amount) {
+        if (amount < 1) {
+            throw new ValidationException(HttpStatus.BAD_REQUEST, "감소 수량은 1 이상이어야 합니다.");
```
도메인 클래스가 http status code를 굳이 알아야 할까요?

seoyoungjin23 replied:
위에서의 답변과 마찬가지로 클라이언트와의 통신에서만 의미가 있는 값이기 때문에 이러한 status코드를 엔티티에 두는 것을 적합하지 못한거 같습니다!

urarik commented on src/main/java/gift/entity/Wish/Wish.java:
```diff
@@ -55,7 +57,7 @@ public Product getProduct() {
-    public Integer getQuantity() {
-        return quantity;
+    public int getQuantity() {
+        return quantity.quantity();
```
quantity.quantity() 가 부자연스럽게 느껴지지 않나요~?

urarik commented on src/main/java/gift/external/KaKaoMessageClient.java:
```diff
@@ -0,0 +1,108 @@
+            .body(body)
+            .retrieve()
+            .onStatus(status -> status.value() == 401, (req, res) -> {
+                throw new UnauthorizedException();
```
여기서 401을 응답받았다고 서버의 클라이언트에게도 401을 응답주면 의미가 달라지지 않을까요? 클라이언트는 401을 받고 다시 인증을 하려고 할텐데 클라이언트가 인증을 잘하면 항상 해당 API의 에러가 해결될 수 있을까요?
KaKaoTokenClient의 응답 코드들도 같이 고려해보세요~

seoyoungjin23 replied:
- 이런 부분은 고려하지 못한거 같습니다! 서버에서 보내는 401이 사용자 입장에서는 자신이 인증을 잘못한 것으로 오해할 여지가 존재하는 것 같습니다! 리팩토링하겠습니다!

urarik commented on src/main/java/gift/service/Order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,51 @@
+        ProductOption option = optionRepository.findById(request.optionId())
+            .orElseThrow(() -> new CustomException(CustomResponseCode.NOT_FOUND));
+
+        option.decreaseQuantity(request.quantity());
```
이 부분에서 동시성 문제가 발생할 수 있습니다. 다른 멘티분에게 드린 댓글인데 한 번 읽어보세요~ https://github.com/next-step/spring-gift-order/pull/197#discussion_r2240172663

다만 **이번 수업 범위를 벗어나는 내용이기 때문에 꼭 다음 PR까지 개선할 필요는 없습니다!** 머릿속에 넣어두시고 나중에 시간될 때 한 번 공부해보세요~

urarik commented on src/test/java/gift/e2e/KaKaoAuthE2ETest.java:
```diff
@@ -44,20 +42,30 @@ void loginSuccessReturnRedirectUrl() throws Exception {
     @Test
-    @DisplayName("콜백 요청시 토큰 반환 성공")
-    void callbackSuccessReturnAccessToken() throws Exception {
+    @DisplayName("콜백 요청 시 쿠키에 액세스 토큰 설정 후 리다이렉트")
+    void callbackSuccessSetAccessTokenInCookie() throws Exception {
```
DisplayName에 테스트의 의도가 다 들어가있으면 메소드 이름을 정하느라 고민하는 것이 불필요할 수도 있지 않을까요~? 영지님은 어떻게 생각하세요?

seoyoungjin23 replied:
완전 동의합니다.. 이 부분도 전체적으로 수정하겠습니다!

urarik reviewed (changes requested):
테스트 코드도 잘 작성해주셨네요~ 코멘트 몇 개 드렸으니 확인해주세요! 한 가지 드리고 싶은 말씀은 조금 오버엔지니어링의 느낌이 납니다ㅎㅎㅎ KISS, YAGNI, DRY 원칙을 한 번 읽어보세요~ 개발을 잘하는 사람은 어려운 문제를 쉽게 푸는 사람이라는 것을 항상 생각하면서 코딩하면 좋아요

질문주신 부분에 대해서, dto의 값은 항상 검증이 성공한다라는 전제를 두면 한 번만 검증해도 될 것 같아요. 중요한건 이 부분을 어떻게 코드상으로 잘 표현하고 강제할 수 있느냐인데요. 이건 영진님이 한 번 고민해보세요! 저라면 굳이 VO를 하나 더 두진 않을 것 같습니다. 제 기준에선 필요 이상으로 복잡해보여요.

seoyoungjin23 committed:
test: DisplayName과 중복되는 테스트 메서드명을 test 등으로 단순화

seoyoungjin23 committed:
refactor: VO 필드의 getter 메서드를 .value()로 변경

- quantity.quantity()와 같은 중복 표현 제거

seoyoungjin23 committed:
refactor: 카카오 API 4xx 오류를 KaKaoClientException으로 통합 처리

기존에는 401, 403, 400 등 클라이언트 오류가 그대로 전파되어
클라이언트가 인증/인가 문제로 오해할 가능성이 있었습니다.

이에 따라 카카오 API 호출 실패 시 발생하는 모든 4xx 오류를
KaKaoClientException으로 통일하여 처리하도록 수정하였습니다!

seoyoungjin23 committed:
fix: 카카오 메시지 전송 시 401 예외만 별도 서버 내부 처리, 나머지 4xx는 공통 예외로 통합

기존에는 메시지 전송 중 발생하는 모든 예외를 내부에서 분기 처리하여 전달했지만 클라이언트가 인증 오류로 오해할 수 있는 문제를 방지하기 위해
401 응답에 대해서만 토큰을 재발급하고 나머지 오류는 공통 예외로 처리하도록 리팩토링하였습니다.

seoyoungjin23 committed:
refactor: CurrentUserArgumentResolver에서 명확한 예외 타입 사용

seoyoungjin23 committed:
refactor: HTTP 상태 코드 설정 방식 표준화

seoyoungjin23 committed:
refactor: 패키지명 소문자로 변경

seoyoungjin23 committed:
refactor: Member 생성 로직을 Builder 패턴으로 변경 및 Member 빌더 사용으로 생성자 호출 제거

seoyoungjin23 committed:
refactor: Member Builder 사용으로 인한 코드 수정

seoyoungjin23 committed:
refactor: Member 객체 생성 방식 new → builder 패턴으로 변경

seoyoungjin23 committed:
refactor: Member 객체 생성 방식 new → builder 패턴으로 변경에 따른 코드 수정

seoyoungjin23 committed:
refactor: 도메인 레이어에서 HttpStatus 의존 제거

seoyoungjin23 committed:
refactor: CustomException에서 응답 코드 책임 제거

seoyoungjin23 committed:
refactor: CustomException에서 응답 코드 책임 제거에 따른 코드 수정

seoyoungjin23 commented:
### 주요 변경 사항

- **카카오 API 4xx 오류를 KaKaoClientException으로 통합 처리**
  - 이 부분에 대해서는 전혀 고려하지 않았었는데 멘토님 덕분에 생각해보게 되었습니다!
  - 카카오 로그인 토큰 클라이언트와 메시지 클라이언트 모두 카카오의 인가 코드를 통해 일차적인 인증 과정을 거친 뒤 혹은 서버에서의 데이터 가공 후에 요청을 수행하게 됩니다. 
  - 이로 인해 발생하는 `4xx` 오류는 일반적으로 클라이언트의 잘못된 요청보다는 서버에서 카카오 API를 사용하는 로직 상의 문제나 잘못된 데이터 가공에서 문제가 있다고 판단했습니다!
  - 따라서 이러한 오류를 클라이언트에게 그대로 전달하기보다는, `400`, `401`, `403` 등의 응답을 서버 내부에서 `KaKaoClientException`으로 감싸 `"외부 API 서버에서 오류가 발생했습니다."`라는 일반화된 메시지를 반환하도록 처리하는 것이 더 적절하다고 판단하여 개선하였습니다!


- **전체적인 예외 처리 개선**
  - 명확한 예외 던지도록 수정
    - 이제 더이상 CustomException이 아닌 명확한 예외를 처리하도록 수정하였습니다!
  - 응답 코드를 예외 처리 핸들러에서 주입
    - 이전과 달리 도메인 혹은 서비스 레이어에서 응답코드를 받지 않습니다! 아무래도 클라이언트와의 통신에서만 의미가 있는 값이기 때문에 이러한 책임은 글로벌 예외 처리기에서 하는 것이 맞다고 판단하였습니다!


- **빌더 만들기**
  - 매번 `lombok`만 사용하다 보니 직접 `builder`를 구현해본 경험이 없었는데 builder를 직접 만드는 과정이 이번 리팩토링에서 가장 재미있고 인상 깊은 경험이었습니다!!


- **어색한 VO 필드 getter 메서드 변경**


- **테스트 메서드명 단순화**


#### 궁금한 사항
- 현재 예외 처리 방식은 `CustomException`을 상속한 예외 클래스 내부에서 HttpStatus를 반환하도록 구성되어 있습니다.
하지만 HttpStatus도 결국 클라이언트 응답하는데만 쓰이는 값이기 때문에 이를 예외 클래스에서 직접 다루는 것이 맞는지에 대해 의문이 생겼습니다.

- 어떤 HTTP 상태로 응답할지는 글로벌 예외 처리기에서 결정하는 것이 더 확실한(?) 책임분리 라고 생각하게 되었고 이러한 면에서 글로벌 예외 처리기에서 예외 타입에 따라 상태 코드와 메시지를 매핑하는 방식이 더 좋은 설계로 느껴졌습니다!

- 위와 같은 방식으로 예외 클래스에서 HttpStatus를 제거하고 글로벌 예외 처리기에서 상태 코드를 분기하거나 예외마다 핸들러를 따로 정의하는 방식으로 리팩토링할 경우 예외가 늘어날수록 글로벌 예외 처리기에 많은 분기문이나 메서드가 생기는 단점이 생길 것 같습니다.

- 그렇다면 이런 구조에서도 명확한 책임 분리를 위해 HttpStatus를 예외에 담지 않는 것이 좋은 선택일까요? 아니면 지금처럼 HttpStatus를 예외 클래스에 포함시켜서 처리하는 구조를 유지하는 것이 더 나은 방향일지가 고민입니다...
 

```
public class ValidationException extends CustomException {

    public ValidationException() {
        super();
    }

    public ValidationException(String message) {
        super(message);
    }

    @Override
    public HttpStatus getStatus() {
        return HttpStatus.BAD_REQUEST;
    }

    @Override
    public String getDefaultMessage() {
        return "유효성 검사에 실패했습니다.";
    }
}
```

seoyoungjin23 commented:
> 테스트 코드도 잘 작성해주셨네요~ 코멘트 몇 개 드렸으니 확인해주세요! 한 가지 드리고 싶은 말씀은 조금 오버엔지니어링의 느낌이 납니다ㅎㅎㅎ KISS, YAGNI, DRY 원칙을 한 번 읽어보세요~ 개발을 잘하는 사람은 어려운 문제를 쉽게 푸는 사람이라는 것을 항상 생각하면서 코딩하면 좋아요
> 
> 질문주신 부분에 대해서, dto의 값은 항상 검증이 성공한다라는 전제를 두면 한 번만 검증해도 될 것 같아요. 중요한건 이 부분을 어떻게 코드상으로 잘 표현하고 강제할 수 있느냐인데요. 이건 영진님이 한 번 고민해보세요! 저라면 굳이 VO를 하나 더 두진 않을 것 같습니다. 제 기준에선 필요 이상으로 복잡해보여요.

감사합니다 멘토님!
말씀 주신 KISS, YAGNI, DRY 원칙들 다시 한번 찾아보며 고민해보겠습니다! 실제로 구현하면서도 이렇게까지 해야하나..? 라는 생각이 자주 들었는데 역시 오버엔지니어링된 것 같습니다ㅎㅎ
지금은 학습이라는 목적이 크기 때문에 일부러라도 여러 시도를 해보는 중입니다. 다만, 말씀해주신 것처럼 "어려운 문제를 쉽게 푸는 사람" 이라는 말이 계속 머릿속에 남습니다... 앞으로는 그 방향성에 대해서 더 신경 써보겠습니다!!

urarik commented on src/test/java/gift/e2e/OrderE2ETest.java:
```diff
@@ -64,12 +65,12 @@ void setUp() {
+            .email("test@domain.com")
+            .nickname("테스트 사용자")
+            .profileImage("https://example.com/profile.jpg")
+            .build();
```
똑같은 멤버 객체 생성이 반복되고 있는 것 같아요~ Test Fixture라는 것을 한 번 찾아보세요!

urarik commented on src/main/java/gift/external/KaKaoMessageClient.java:
```diff
@@ -39,12 +41,27 @@ public KaKaoMessageClient(RestClient.Builder restClientBuilder, ObjectMapper obj
+            }
+            throw new KaKaoClientException(CustomResponseCode.KAKAO_MESSAGE_SEND_FAILED);
+        } catch (HttpServerErrorException e) {
+            throw new ServerErrorException(CustomResponseCode.SERVER_ERROR);
```
자바에도 몇몇 표준 예외를 제공해요. effective java item 72 를 검색해서 관련 글을 읽어보면 좋을 것 같습니다~
반드시 표준 예외만 사용하라는 것은 아니고 이러한 관점도 있다는 것을 알아두면 좋을 것 같아서 코멘트 남깁니다!

urarik reviewed (approved):
수고 많으셨습니다~ 제 코멘트를 잘 받아들여주셔서 저도 감사하네요 ㅎㅎ 머지할게요~~

 궁금한 점에 대해선, 코멘트에서도 달았듯이 예외가 발생했을 때 항상 예외 클래스를 만드는 것보다 표준 예외를 많이 활용해보면 좋아요. 그럼 예외 클래스가 많이 늘어나진 않을거예요. 그리고 응답 코드에 대해서도 너무 엄격하게 지키려고 하지 않으셔도 됩니다. 응답 코드가 달라짐에 따라 클라이언트가 행동을 다르게 해야하는 경우에는 응답 코드가 중요하겠지만 실용적으로 봤을 때 대부분의 케이스에선 200, 400, 404, 500 정도만 내려줘도 충분하거든요. 이 케이스는 정답이 있는 것이 아니기 때문에 영진님이 방식들의 장단점을 잘 생각해보고 결정해보세요!

 영진님이 지금 읽기엔 내용이 꽤나 어려울텐데 이번 과정이 끝나고 추후에 시간이 날 때 effective java라는 책을 읽어보면 좋을 것 같아요. 자바에 대한 부분뿐만 아니라 외부에 공개되는 API를 어떻게 설계하느냐에 대한 전반적인 부분을 많이 설명해주기 때문에 저에겐 많은 도움이 된 책이었어요.

 그리고 책임 분리를 왜하는지에 대한 근본적인 이유들도 많이 생각해보면 좋을 것 같아요. 기계적으로 책임 분리를 하면 좋다는 생각만 가지고 하려다보면 판단을 내리기 되게 모호할 때가 많아요. 책임 이라는 것이 매우 주관적인 판단이 들어가는 부분이기도 하고 책임 분리과 유지보수성 등 다른 가치들이 충돌을 일으킬 때가 있거든요. 그럴 때 한 가지 방법을 선택하기 위해 만들 수 있는 근거들을 세우기 위해선 각각의 방법들이 왜 필요한지에 대한 근본적인 생각, 실용주의적 관점 등이 모두 필요한 것 같아요.

영진님이 잘 반영해주셔서 말이 길어졌네요 ㅎㅎ 어쨌든 잘 작성해주셨습니다 수고하셨습니다~
